c7n/resources/cw.py

```diff
@@ -114,6 +119,135 @@ def get_dimensions(self, resource):
         return [{'Name': 'RuleName', 'Value': resource['Name']}]
 
 
+@EventRule.filter_registry.register('event-rule-target')
+class EventRuleTargetFilter(ChildResourceFilter):
+    """
+    Filter event rules by their targets
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+            - name: find-event-rules-with-no-targets
+              resource: aws.event-rule
+              filters:
+                - type: event-rule-target
+                  key: Arn
+                  value: absent
+    """
+
+    RelatedResource = "c7n.resources.cw.EventRuleTarget"
+    RelatedIdsExpression = 'Name'
+    AnnotationKey = "EventRuleTargets"
+
+    schema = type_schema('event-rule-target', rinherit=ValueFilter.schema)
+    permissions = ('events:ListTargetsByRule',)
+
+
+@EventRule.filter_registry.register('invalid-targets')
+class ValidEventRuleTargetFilter(ChildResourceFilter):
+    """
+    Filter event rules for invalid targets
+
+    :example:
+
+    .. code-block:: yaml
+
+        policies:
+            - name: find-event-rules-with-invalid-targets
+              resource: aws.event-rule
+              filters:
+                - type: invalid-targets
+    """
+
+    RelatedResource = "c7n.resources.cw.EventRuleTarget"
+    RelatedIdsExpression = 'Name'
+    AnnotationKey = "EventRuleTargets"
+
+    schema = type_schema('invalid-targets',)
+    permissions = ('events:ListTargetsByRule',)
+
+    def _get_arns(self, resources):
+        arn_map = {}
+        r_map = {}
+        child_resources = self.get_related(resources)
+        for r in resources:
+            r_map[r['Name']] = r
+        for k, v in child_resources.items():
+            for r in v:
+                arn_map.setdefault(k, {}).setdefault('Arns', []).append(r['Arn'])
+                arn_map[k]['data'] = r_map[v[0]['c7n:parent-id']]
+        return arn_map
+
+    def process(self, resources, event=None):
+        # Due to lazy loading of resources, we need to explicilty load the following
+        # potential targets for a event rule target:
+        load_resources(
+            [
+                "aws.sqs",
+                "aws.lambda",
+                "aws.ecs-cluster",
+                "aws.ecs-task",
+                "aws.kinesis",
+                "aws.sns",
+                "aws.ssm-parameter",
+                "aws.batch-compute",
+                "aws.codepipeline",
+            ]
+        )
+        arn_resolver = ArnResolver(self.manager)
+        arn_map = self._get_arns(resources)
+        results = []
+        for k, v in arn_map.items():
+            resolved = arn_resolver.resolve(v['Arns'])
+            if not all(resolved.values()):
+                for i, j in resolved.items():
+                    if not j:
+                        v['data'].setdefault('c7n:InvalidTargets', []).append(i)
+                results.append(v['data'])
```

Comment:
is this transforming resources to resource-ids? tracing the data through the get_arns arn_map is a little tricky, might be useful to document the dict structure ... arn is pretty ambigious here, going across parent/child and target arns.

#  key: target arn  value: data: rule, target id

Reply:
yea agreed, there was a bit of book keeping that had to be done to properly map the parent/child relationships back for the first pass, I'll try to simply it

Reply:
simplified it significantly, now it's just annotating the original resource dict with the child arns